pkg/daemon/addr/addr.go

```diff
@@ -0,0 +1,214 @@
+	interfaceToSubnetMap map[string]subnetToPodMap
+}
+
+func CreateAddrManager(family int, nodeName string) *Manager {
+	return &Manager{
+		family:               family,
+		localNodeName:        nodeName,
+		interfaceToSubnetMap: map[string]subnetToPodMap{},
+	}
+}
+
+func (m *Manager) ResetInfos() {
+	m.interfaceToSubnetMap = map[string]subnetToPodMap{}
+}
+
+func (m *Manager) TryAddPodInfo(forwardNodeIfName string, subnet *net.IPNet, podIP net.IP) {
+	if subnetMap := m.interfaceToSubnetMap[forwardNodeIfName]; subnetMap == nil {
+		m.interfaceToSubnetMap[forwardNodeIfName] = subnetToPodMap{}
+	}
+
+	// we only need one local pod ip for every subnet
+	if _, exist := m.interfaceToSubnetMap[forwardNodeIfName][subnet.String()]; !exist {
+		m.interfaceToSubnetMap[forwardNodeIfName][subnet.String()] = podIP
+	}
+}
+
+// For some environments, physical router or switcher might check the sender address
+// of arp request, if the sender ip address is not in the same subnet of target address
+// the arp request will be take as invalid and dropped.
+//
+// So we will always keep an valid local pod address in the vlan interface without local routes.
+func (m *Manager) SyncAddresses(getIPInstanceByAddress func(net.IP) (*ramav1.IPInstance, error)) error {
+	// clear all invalid enhanced addresses
+	linkList, err := netlink.LinkList()
+	if err != nil {
+		return fmt.Errorf("list link failed: %v", err)
+	}
+
+	existEnhancedAddrMap := map[string]map[string]netlink.Addr{}
+	existManualAddrSubnetMap := map[string]map[string]bool{}
+
+	for _, link := range linkList {
+		// ignore container network virtual interfaces
+		if strings.HasSuffix(link.Attrs().Name, containernetwork.ContainerHostLinkSuffix) ||
+			strings.HasSuffix(link.Attrs().Name, containernetwork.ContainerInitLinkSuffix) ||
+			strings.HasPrefix(link.Attrs().Name, "veth") {
+
+			continue
+		}
+
+		addrList, err := netlink.AddrList(link, m.family)
+		if err != nil {
+			return fmt.Errorf("list addresses for link %v failed: %v", link.Attrs().Name, err)
+		}
+
+		for _, addr := range addrList {
+			isEnhancedAddr, err := checkIfEnhancedAddr(link, addr, m.family)
+			if err != nil {
+				return fmt.Errorf("check addr %v enhanced address failed: %v", addr.String(), err)
+			}
+
+			linkName := link.Attrs().Name
+			cidr := net.IPNet{
+				IP:   addr.IP.Mask(addr.Mask),
+				Mask: addr.Mask,
+			}
+
+			if isEnhancedAddr {
+				if existEnhancedAddrMap[linkName] == nil {
+					existEnhancedAddrMap[linkName] = map[string]netlink.Addr{}
+				}
+				existEnhancedAddrMap[linkName][cidr.String()] = addr
+			} else {
+				if existManualAddrSubnetMap[linkName] == nil {
+					existManualAddrSubnetMap[linkName] = map[string]bool{}
+				}
+				existManualAddrSubnetMap[linkName][cidr.String()] = true
+			}
+		}
+	}
+
+	// clear enhanced addresses which are impossible to be used
+	for existLinkName, existSubnetMap := range existEnhancedAddrMap {
+		existLink, err := netlink.LinkByName(existLinkName)
```

Comment:
This link is just listed before, so no need to check it existence, right?

Reply:
fixed~